File: JeMPI_Apps/JeMPI_Configuration/src/main/scala/configuration/CustomDgraphQueries.scala
package configuration

import java.io.{File, PrintWriter}
import scala.language.{existentials, postfixOps}


object CustomDgraphQueries {

  private val classLocation = "../JeMPI_LibMPI/src/main/java/org/jembi/jempi/libmpi/dgraph"
  private val custom_className = "CustomDgraphQueries"
  private val packageText = "org.jembi.jempi.libmpi.dgraph"

  def generate(config: Config): Any = {

    val classFile: String = classLocation + File.separator + custom_className + ".java"
    println("Creating " + classFile)
    val file: File = new File(classFile)
    val writer: PrintWriter = new PrintWriter(file)
    writer.println(
      s"""package $packageText;
         |
         |import io.vavr.Function1;
         |import org.apache.commons.lang3.StringUtils;
         |import org.jembi.jempi.shared.models.CustomDemographicData;
         |
         |import java.util.LinkedList;
         |import java.util.List;
         |import java.util.Map;
         |
         |import static org.jembi.jempi.libmpi.dgraph.DgraphQueries.runGoldenRecordsQuery;
         |
         |final class $custom_className {
         |
         |   static final List<Function1<CustomDemographicData, DgraphGoldenRecords>> DETERMINISTIC_LINK_FUNCTIONS =
         |      List.of(${getDeterministicFunctions(config.rules.link.get)});
         |""".stripMargin)

    if (config.rules.matchNotification.isDefined) {
      writer.println(
        s"""   static final List<Function1<CustomDemographicData, DgraphGoldenRecords>> DETERMINISTIC_MATCH_FUNCTIONS =
           |      List.of(${getDeterministicFunctions(config.rules.matchNotification.get)});
           |""".stripMargin)
    } else {
      writer.println(
        s"""   static final List<Function1<CustomDemographicData, DgraphGoldenRecords>> DETERMINISTIC_MATCH_FUNCTIONS =
           |      List.of();
           |""".stripMargin)
    }

    if (config.rules.link.get.deterministic.isDefined) {
      config.rules.link.get.deterministic.get.foreach((name, rule) => emitRuleTemplate(name, rule))
    }
    if (config.rules.link.isDefined && config.rules.link.get.probabilistic.isDefined) {
      config.rules.link.get.probabilistic.get.foreach((name, rule) => emitRuleTemplate(name, rule))
    }

    if (config.rules.matchNotification.isDefined && config.rules.matchNotification.get.deterministic.isDefined) {
      config.rules.matchNotification.get.deterministic.get.foreach((name, rule) => emitRuleTemplate(name, rule))
    }
    if (config.rules.matchNotification.isDefined && config.rules.matchNotification.get.probabilistic.isDefined) {
      config.rules.matchNotification.get.probabilistic.get.foreach((name, rule) => emitRuleTemplate(name, rule))
    }

    if (config.rules.link.isDefined && config.rules.link.get.deterministic.isDefined) {
      config.rules.link.get.deterministic.get.foreach((name, rule) => emitRuleFunction(name, rule))
    }
    if (config.rules.link.isDefined && config.rules.link.get.probabilistic.isDefined) {
      config.rules.link.get.probabilistic.get.foreach((name, rule) => emitRuleFunction(name, rule))
    }

    emitMergeCandidates()

    emitFindCandidates("Link", "DETERMINISTIC_LINK_FUNCTIONS", config.rules.link)

    if (config.rules.matchNotification.isDefined && config.rules.matchNotification.get.deterministic.isDefined) {
      config.rules.matchNotification.get.deterministic.get.foreach((name, rule) => emitRuleFunction(name, rule))
    }
    if (config.rules.matchNotification.isDefined && config.rules.matchNotification.get.probabilistic.isDefined && config.rules.matchNotification.get.probabilistic.isDefined) {
      config.rules.matchNotification.get.probabilistic.get.foreach((name, rule) => emitRuleFunction(name, rule))
    }
    emitFindCandidates("Match", "DETERMINISTIC_MATCH_FUNCTIONS", config.rules.matchNotification)

    writer.println(
      s"""   private $custom_className() {
         |   }
         |
         |}""".stripMargin)
    writer.flush()
    writer.close()


    def getDeterministicFunctions(allRules: AllRules): String = {
      allRules.deterministic.get
        .map((name, _) => s"""${" " * 14}CustomDgraphQueries::${Utils.snakeCaseToCamelCase(name.toLowerCase)},""".stripMargin)
        .mkString("\n")
        .trim
        .dropRight(1)
    }

    def emitFindCandidates(funcQualifier: String, filterList: String, rules: Option[AllRules]): Unit = {
      writer.println(
        s"""   static List<CustomDgraphGoldenRecord> find${funcQualifier}Candidates(
           |      final CustomDemographicData interaction) {
           |      var result = DgraphQueries.deterministicFilter($filterList, interaction);
           |      if (!result.isEmpty()) {
           |         return result;
           |      }
           |      result = new LinkedList<>();""".stripMargin)
      if (rules.isDefined && rules.get.probabilistic.isDefined) {
        rules.get.probabilistic.get.foreach((name, _) => {
          val filterName = Utils.snakeCaseToCamelCase(name.toLowerCase)
          val vars = "interaction"
          writer.println(s"""${" " * 6}mergeCandidates(result, $filterName($vars));""".stripMargin)
        })
      }
      writer.println(
        """      return result;
          |   }
          |""".stripMargin)
    }

    def emitRuleFunction(name: String, rule: Rule): Unit = {

      def expression(expr: Ast.Expression): String = {
        expr match {
          case Ast.Or(x) =>
            "("
              + (for (k <- x.zipWithIndex) yield if (k._2 == 0) expression(k._1) else " && " + expression(k._1)).mkString
              + ")"
          case Ast.And(x) =>
            "("
              + (for (k <- x.zipWithIndex) yield if (k._2 == 0) expression(k._1) else " || " + expression(k._1)).mkString
              + ")"
          case Ast.Not(x) =>
            "NOT (" + expression(x) + ")"
          case Ast.Match(variable, _) =>
            Utils.snakeCaseToCamelCase(variable.name) + "IsBlank"
          case Ast.Eq(variable) =>
            Utils.snakeCaseToCamelCase(variable.name) + "IsBlank"
          case _ =>
            "ERROR"
        }
      }

      val vars = for (v <- rule.vars) yield v
      val text = rule.text

      val functionName = Utils.snakeCaseToCamelCase(name.toLowerCase)
      if (vars.length == 1)
        val v = vars(0)
        writer.println(
          s"""   private static DgraphGoldenRecords $functionName(final CustomDemographicData demographicData) {
             |      if (StringUtils.isBlank(demographicData.${Utils.snakeCaseToCamelCase(v)})) {
             |         return new DgraphGoldenRecords(List.of());
             |      }
             |      final Map<String, String> map = Map.of("$$$v", demographicData.${Utils.snakeCaseToCamelCase(v)});
             |      return runGoldenRecordsQuery($name, map);
             |   }
             |""".stripMargin)
      else
        val expr = expression(ParseRule.parse(text))
        writer.println(s"   private static DgraphGoldenRecords $functionName(final CustomDemographicData demographicData) {")
        vars.foreach(v => {
          val camelCaseVarName = Utils.snakeCaseToCamelCase(v)
          writer.println(s"      final var $camelCaseVarName = demographicData.$camelCaseVarName;")
        })
        vars.foreach(v => {
          val camelCaseVarName = Utils.snakeCaseToCamelCase(v)
          val isNullVar = camelCaseVarName + s"IsBlank"
          writer.println(s"      final var $isNullVar = StringUtils.isBlank($camelCaseVarName);")
        })
        writer.print(
          s"""      if ($expr) {
             |         return new DgraphGoldenRecords(List.of());
             |      }
             |      final var map = Map.of(""".stripMargin)

        vars.zipWithIndex.foreach((v, idx) => {
          val camelCaseVarName = Utils.snakeCaseToCamelCase(v)
          writer.println(
            s"""${" " * (if (idx == 0) 0 else 29)}"$$$v",
               |${" " * 29}StringUtils.isNotBlank($camelCaseVarName)
               |${" " * 29}      ? $camelCaseVarName
               |${" " * 29}      : DgraphQueries.EMPTY_FIELD_SENTINEL${if (idx + 1 < vars.length) "," else ");"}""".stripMargin)
        })
        writer.println(
          s"""      return runGoldenRecordsQuery($name, map);
             |   }
             |""".stripMargin)
      end if
    }

    def emitRuleTemplate(name: String, rule: Rule): Unit = {

      val vars = for (v <- rule.vars) yield v
      val text = rule.text
      val expression: Ast.Expression = ParseRule.parse(text)
      val varsMap = vars.zipWithIndex.toMap[String, Int].map((k, i) => (k, ("A".head + i).toChar.toString))
      var meta = Map[String, (String, Option[Integer])]()

      def main_func(expression: Ast.Expression): String = {
        expression match {
          case Ast.Or(x) => "("
            + (for (k <- x.zipWithIndex) yield if (k._2 == 0) main_func(k._1) else " OR " + main_func(k._1)).mkString
            + ")"
          case Ast.And(x) => "("
            + (for (k <- x.zipWithIndex) yield if (k._2 == 0) main_func(k._1) else " AND " + main_func(k._1)).mkString
            + ")"
          case Ast.Not(x) =>
            "NOT (" + main_func(x) + ")"
          case Ast.Match(variable, distance) =>
            meta += (variable.name -> ("match", Option(distance)))
            "uid(" + variable.name + ")"
          case Ast.Eq(variable) =>
            meta += (variable.name -> ("eq", None))
            "uid(" + variable.name + ")"
          case _ =>
            "ERROR"
        }
      }

      def createScalerFunc(): Unit = {
        vars.foreach(v => {
          val fn = meta(v)._1
          writer.println(
            s"""${" " * 12}all(func:type(GoldenRecord)) @filter($fn(GoldenRecord.$v, $$$v${
              if (meta(v)._2.isDefined) ", " + meta(v)._2.get else
                ""
            })) {
               |${" " * 15}uid
               |${" " * 15}GoldenRecord.source_id {
               |${" " * 18}uid
               |${" " * 15}}""".stripMargin)
          if (config.uniqueGoldenRecordFields.isDefined) {
            config.uniqueGoldenRecordFields.get.foreach(field => {
              writer.println(s"${" " * 15}GoldenRecord.${field.fieldName}")
            })
          }
          config.demographicFields.foreach(field => {
            writer.println(s"${" " * 15}GoldenRecord.${field.fieldName}")
          })
          writer.println(s"${" " * 12}}")
        })
      }

      def createFilterFunc(all_func_str: String): Unit = {
        vars.foreach(v => {
          val fn = meta(v)._1
          writer.println(
            s"""${" " * 12}var(func:type(GoldenRecord)) @filter($fn(GoldenRecord.$v, $$$v${
              if (meta(v)._2.isDefined) ", " + meta(v)._2.get else ""
            })) {
               |${" " * 15}${varsMap(v)} as uid
               |${" " * 12}}""".stripMargin)
        })
        writer.println(
          s"""${" " * 12}all(func:type(GoldenRecord)) @filter${
            if (all_func_str.startsWith("(")) "" else "("
          }$all_func_str${if (all_func_str.startsWith("(")) "" else "("} {
             |${" " * 15}uid
             |${" " * 15}GoldenRecord.source_id {
             |${" " * 18}uid
             |${" " * 15}}""".stripMargin)
        if (config.uniqueGoldenRecordFields.isDefined) {
          config.uniqueGoldenRecordFields.get.foreach(field => {
            writer.println(s"${" " * 15}GoldenRecord.${field.fieldName}")
          })
        }
        config.demographicFields.foreach(field => {
          writer.println(s"${" " * 15}GoldenRecord.${field.fieldName}")
        })
        writer.println(s"${" " * 12}}")
      }

      var all_func_str = main_func(expression)
      varsMap.foreach((k, v) => {
        all_func_str = all_func_str.replace("uid(" + k + ")", "uid(" + v + ")")
      })

      writer.print(
        s"""${" " * 3}private static final String ${name.toUpperCase} =
           |${" " * 9}\"\"\"
           |${" " * 9}query ${name.toLowerCase}(""".stripMargin)

      vars.zipWithIndex.foreach((v, i) => {
        if (i > 0)
          writer.print(", ")
        end if
        writer.print(s"$$$v: string")
      })
      writer.println(") {")

      if (vars.length == 1)
        createScalerFunc()
      else
        createFilterFunc(all_func_str)
      end if
      writer.println(
        s"""${" " * 9}}
           |${" " * 9}\"\"\";
           |""".stripMargin)
    }

    def emitMergeCandidates(): Unit = {
      writer.println(
        s"""   private static void mergeCandidates(
           |         final List<CustomDgraphGoldenRecord> goldenRecords,
           |         final DgraphGoldenRecords block) {
           |      final var candidates = block.all();
           |      if (!candidates.isEmpty()) {
           |         candidates.forEach(candidate -> {
           |            var found = false;
           |            for (CustomDgraphGoldenRecord goldenRecord : goldenRecords) {
           |               if (candidate.goldenId().equals(goldenRecord.goldenId())) {
           |                  found = true;
           |                  break;
           |               }
           |            }
           |            if (!found) {
           |               goldenRecords.add(candidate);
           |            }
           |         });
           |      }
           |   }
           |""".stripMargin)
    }

  }

}

File: JeMPI_Apps/JeMPI_LibConfig/src/main/java/org/jembi/jempi/libconfig/dgraph/CustomDgraphQueries.java
package org.jembi.jempi.libconfig.dgraph;

import org.apache.commons.lang3.StringUtils;
import org.jembi.jempi.libconfig.shared.models.CustomDemographicData;
import org.jembi.jempi.libconfig.shared.utils.AppUtils;

import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;

import static org.jembi.jempi.libconfig.shared.models.CustomDemographicData.*;

public final class CustomDgraphQueries {

   static final String EMPTY_FIELD_SENTINEL = "EMPTY_FIELD_SENTINEL";
   static final List<
         BiFunction<BiFunction<String, Map<String, String>, DgraphGoldenRecords>,
               CustomDemographicData,
               DgraphGoldenRecords>> DETERMINISTIC_MATCH_FUNCTIONS =
         List.of();

   static final List<
         BiFunction<
               BiFunction<
                     String, Map<String, String>, DgraphGoldenRecords>,
               CustomDemographicData,
               DgraphGoldenRecords>> DETERMINISTIC_LINK_FUNCTIONS =
         List.of(CustomDgraphQueries::queryLinkDeterministicA,
                 CustomDgraphQueries::queryLinkDeterministicB);

   private static final String QUERY_LINK_DETERMINISTIC_A =
         """
         query query_link_deterministic_a($national_id: string) {
            all(func:type(GoldenRecord)) @filter(eq(GoldenRecord.national_id, $national_id)) {
               uid
               GoldenRecord.source_id {
                  uid
               }
               GoldenRecord.aux_date_created
               GoldenRecord.aux_auto_update_enabled
               GoldenRecord.aux_id
               GoldenRecord.given_name
               GoldenRecord.family_name
               GoldenRecord.gender
               GoldenRecord.dob
               GoldenRecord.city
               GoldenRecord.phone_number
               GoldenRecord.national_id
            }
         }
         """;
   private static final String QUERY_LINK_DETERMINISTIC_B =
         """
         query query_link_deterministic_b($given_name: string, $family_name: string, $phone_number: string) {
            var(func:type(GoldenRecord)) @filter(eq(GoldenRecord.given_name, $given_name)) {
               A as uid
            }
            var(func:type(GoldenRecord)) @filter(eq(GoldenRecord.family_name, $family_name)) {
               B as uid
            }
            var(func:type(GoldenRecord)) @filter(eq(GoldenRecord.phone_number, $phone_number)) {
               C as uid
            }
            all(func:type(GoldenRecord)) @filter(uid(A) AND uid(B) AND uid(C)) {
               uid
               GoldenRecord.source_id {
                  uid
               }
               GoldenRecord.aux_date_created
               GoldenRecord.aux_auto_update_enabled
               GoldenRecord.aux_id
               GoldenRecord.given_name
               GoldenRecord.family_name
               GoldenRecord.gender
               GoldenRecord.dob
               GoldenRecord.city
               GoldenRecord.phone_number
               GoldenRecord.national_id
            }
         }
         """;
   private static final String QUERY_LINK_PROBABILISTIC =
         """
         query query_link_probabilistic($given_name: string, $family_name: string, $city: string, $phone_number: string, $national_id: string) {
            var(func:type(GoldenRecord)) @filter(match(GoldenRecord.given_name, $given_name, 3)) {
               A as uid
            }
            var(func:type(GoldenRecord)) @filter(match(GoldenRecord.family_name, $family_name, 3)) {
               B as uid
            }
            var(func:type(GoldenRecord)) @filter(match(GoldenRecord.city, $city, 3)) {
               C as uid
            }
            var(func:type(GoldenRecord)) @filter(match(GoldenRecord.phone_number, $phone_number, 2)) {
               D as uid
            }
            var(func:type(GoldenRecord)) @filter(match(GoldenRecord.national_id, $national_id, 3)) {
               E as uid
            }
            all(func:type(GoldenRecord)) @filter(((uid(A) AND uid(B)) OR (uid(A) AND uid(C)) OR (uid(B) AND uid(C))) OR uid(D) OR uid(E)) {
               uid
               GoldenRecord.source_id {
                  uid
               }
               GoldenRecord.aux_date_created
               GoldenRecord.aux_auto_update_enabled
               GoldenRecord.aux_id
               GoldenRecord.given_name
               GoldenRecord.family_name
               GoldenRecord.gender
               GoldenRecord.dob
               GoldenRecord.city
               GoldenRecord.phone_number
               GoldenRecord.national_id
            }
         }
         """;

   private CustomDgraphQueries() {
   }

   private static DgraphGoldenRecords queryLinkDeterministicA(
         final BiFunction<String, Map<String, String>, DgraphGoldenRecords> runGoldenRecordsQuery,
         final CustomDemographicData demographicData) {
      if (StringUtils.isBlank(demographicData.fields[NATIONAL_ID_IDX].value())) {
         return new DgraphGoldenRecords(List.of());
      }
      final Map<String, String> map = Map.of("$national_id", demographicData.fields[NATIONAL_ID_IDX].value());
      return runGoldenRecordsQuery.apply(QUERY_LINK_DETERMINISTIC_A, map);
   }

   private static DgraphGoldenRecords queryLinkDeterministicB(
         final BiFunction<String, Map<String, String>, DgraphGoldenRecords> runGoldenRecordsQuery,
         final CustomDemographicData demographicData) {
      final var givenName = demographicData.fields[GIVEN_NAME_IDX].value();
      final var familyName = demographicData.fields[FAMILY_NAME_IDX].value();
      final var phoneNumber = demographicData.fields[PHONE_NUMBER_IDX].value();
      final var givenNameIsBlank = StringUtils.isBlank(givenName);
      final var familyNameIsBlank = StringUtils.isBlank(familyName);
      final var phoneNumberIsBlank = StringUtils.isBlank(phoneNumber);
      if ((givenNameIsBlank || familyNameIsBlank || phoneNumberIsBlank)) {
         return new DgraphGoldenRecords(List.of());
      }
      final var map = Map.of("$given_name",
                             StringUtils.isNotBlank(givenName)
                                   ? givenName
                                   : EMPTY_FIELD_SENTINEL,
                             "$family_name",
                             StringUtils.isNotBlank(familyName)
                                   ? familyName
                                   : EMPTY_FIELD_SENTINEL,
                             "$phone_number",
                             StringUtils.isNotBlank(phoneNumber)
                                   ? phoneNumber
                                   : EMPTY_FIELD_SENTINEL);
      return runGoldenRecordsQuery.apply(QUERY_LINK_DETERMINISTIC_B, map);
   }

   private static DgraphGoldenRecords queryLinkProbabilistic(
         final BiFunction<String, Map<String, String>, DgraphGoldenRecords> runGoldenRecordsQuery,
         final CustomDemographicData demographicData) {
      final var givenName = demographicData.fields[GIVEN_NAME_IDX].value();
      final var familyName = demographicData.fields[FAMILY_NAME_IDX].value();
      final var city = demographicData.fields[CITY_IDX].value();
      final var phoneNumber = demographicData.fields[PHONE_NUMBER_IDX].value();
      final var nationalId = demographicData.fields[NATIONAL_ID_IDX].value();
      final var givenNameIsBlank = StringUtils.isBlank(givenName);
      final var familyNameIsBlank = StringUtils.isBlank(familyName);
      final var cityIsBlank = StringUtils.isBlank(city);
      final var phoneNumberIsBlank = StringUtils.isBlank(phoneNumber);
      final var nationalIdIsBlank = StringUtils.isBlank(nationalId);
      if ((((givenNameIsBlank || familyNameIsBlank) && (givenNameIsBlank || cityIsBlank) && (familyNameIsBlank || cityIsBlank)) && phoneNumberIsBlank && nationalIdIsBlank)) {
         return new DgraphGoldenRecords(List.of());
      }
      final var map = Map.of("$given_name",
                             StringUtils.isNotBlank(givenName)
                                   ? givenName
                                   : EMPTY_FIELD_SENTINEL,
                             "$family_name",
                             StringUtils.isNotBlank(familyName)
                                   ? familyName
                                   : EMPTY_FIELD_SENTINEL,
                             "$city",
                             StringUtils.isNotBlank(city)
                                   ? city
                                   : EMPTY_FIELD_SENTINEL,
                             "$phone_number",
                             StringUtils.isNotBlank(phoneNumber)
                                   ? phoneNumber
                                   : EMPTY_FIELD_SENTINEL,
                             "$national_id",
                             StringUtils.isNotBlank(nationalId)
                                   ? nationalId
                                   : EMPTY_FIELD_SENTINEL);
      return runGoldenRecordsQuery.apply(QUERY_LINK_PROBABILISTIC, map);
   }

   private static void mergeCandidates(
         final List<CustomDgraphGoldenRecord> goldenRecords,
         final DgraphGoldenRecords block) {
      final var candidates = block.all();
      if (!candidates.isEmpty()) {
         candidates.forEach(candidate -> {
            var found = false;
            for (CustomDgraphGoldenRecord goldenRecord : goldenRecords) {
               if (candidate.goldenId().equals(goldenRecord.goldenId())) {
                  found = true;
                  break;
               }
            }
            if (!found) {
               goldenRecords.add(candidate);
            }
         });
      }
   }

   static LinkedList<CustomDgraphGoldenRecord> deterministicFilter(
         final BiFunction<String, Map<String, String>, DgraphGoldenRecords> runQuery,
         final List<
               BiFunction<
                     BiFunction<
                           String,
                           Map<String, String>,
                           DgraphGoldenRecords>,
                     CustomDemographicData,
                     DgraphGoldenRecords>> listFunction,
         final CustomDemographicData interaction) {
      final LinkedList<CustomDgraphGoldenRecord> candidateGoldenRecords = new LinkedList<>();
      for (BiFunction<
            BiFunction<
                  String, Map<String, String>, DgraphGoldenRecords>,
            CustomDemographicData,
            DgraphGoldenRecords> deterministicFunction : listFunction) {
         final var block = deterministicFunction.apply(runQuery, interaction);
         if (!block.all().isEmpty()) {
            final var list = block.all();
            if (!AppUtils.isNullOrEmpty(list)) {
               candidateGoldenRecords.addAll(list);
               return candidateGoldenRecords;
            }
         }
      }
      return candidateGoldenRecords;
   }

   public static List<CustomDgraphGoldenRecord> findLinkCandidates(
         final BiFunction<String, Map<String, String>, DgraphGoldenRecords> runQuery,
         final CustomDemographicData interaction) {
      var result = deterministicFilter(runQuery, DETERMINISTIC_LINK_FUNCTIONS, interaction);
      if (!result.isEmpty()) {
         return result;
      }
      result = new LinkedList<>();
      mergeCandidates(result, queryLinkProbabilistic(runQuery, interaction));
      return result;
   }

   public static List<CustomDgraphGoldenRecord> findMatchCandidates(
         final BiFunction<String, Map<String, String>, DgraphGoldenRecords> runQuery,
         final CustomDemographicData interaction) {
      var result = deterministicFilter(runQuery, DETERMINISTIC_MATCH_FUNCTIONS, interaction);
      if (!result.isEmpty()) {
         return result;
      }
      result = new LinkedList<>();
      return result;
   }


}
